edb_stat_statements/edb_stat_statements.c

```diff
@@ -0,0 +1,219 @@
+    int nested_level;
+    EdbStmtInfoParseState state;
+    char *query;
+    uint64 queryId;
+} EdbStmtInfoSemState;
+
+static post_parse_analyze_hook_type prev_post_parse_analyze_hook = NULL;
+static planner_hook_type prev_planner_hook = NULL;
+static ExecutorEnd_hook_type prev_ExecutorEnd = NULL;
+
+static void
+edbss_post_parse_analyze(ParseState *pstate, Query *query, JumbleState *jstate);
+
+static PlannedStmt *
+edbss_planner(Query *parse,
+              const char *query_string,
+              int cursorOptions,
+              ParamListInfo boundParams);
+
+static void
+edbss_ExecutorEnd(QueryDesc *queryDesc);
+
+static uint64
+edbss_extract_stmt_info(char **query_str);
+
+static JsonParseErrorType
+edbss_json_struct_start(void *semstate);
+
+static JsonParseErrorType
+edbss_json_struct_end(void *semstate);
+
+static JsonParseErrorType
+edbss_json_ofield_start(void *semstate, char *fname, bool isnull);
+
+static JsonParseErrorType
+edbss_json_scalar(void *semstate, char *token, JsonTokenType tokenType);
+
+void
+_PG_init(void) {
+    prev_post_parse_analyze_hook = post_parse_analyze_hook;
+    post_parse_analyze_hook = edbss_post_parse_analyze;
+    prev_planner_hook = planner_hook;
+    planner_hook = edbss_planner;
+    prev_ExecutorEnd = ExecutorEnd_hook;
+    ExecutorEnd_hook = edbss_ExecutorEnd;
+}
+
+/*
+ * Post-parse-analysis hook: mark query with custom queryId
+ */
+static void
+edbss_post_parse_analyze(ParseState *pstate, Query *query, JumbleState *jstate) {
+    // Parse EdgeDB query info JSON and overwrite query->queryId
+    char *query_str = (char *) pstate->p_sourcetext;
+    query->queryId = edbss_extract_stmt_info(&query_str);
+
+    // But skip pgss_store() as we don't need the early entry
+    if (prev_post_parse_analyze_hook)
+        prev_post_parse_analyze_hook(pstate, query, NULL);
+}
+
+static PlannedStmt *
+edbss_planner(Query *parse,
+              const char *query_string,
+              int cursorOptions,
+              ParamListInfo boundParams) {
+    char *query_str = (char *) query_string;
+    edbss_extract_stmt_info(&query_str);
+    if (prev_planner_hook)
+        return prev_planner_hook(parse, query_str, cursorOptions,
+                                 boundParams);
+    else
+        return standard_planner(parse, query_str, cursorOptions,
+                                boundParams);
+}
+
+static void
+edbss_ExecutorEnd(QueryDesc *queryDesc) {
+    const char *orig_sourceText = queryDesc->sourceText;
+    edbss_extract_stmt_info((char **) &queryDesc->sourceText);
+    if (prev_ExecutorEnd)
+        prev_ExecutorEnd(queryDesc);
+    else
+        standard_ExecutorEnd(queryDesc);
+    queryDesc->sourceText = orig_sourceText;
+}
+
+/*
+ * Extract EdgeDB statement info and overwrite source text and queryId
+ */
+static uint64
+edbss_extract_stmt_info(char **query_str) {
+    if (strncmp(*query_str, EDB_STMT_MAGIC_PREFIX, strlen(EDB_STMT_MAGIC_PREFIX)) == 0) {
+        EdbStmtInfoSemState state;
```

Comment:
What version of C will this be compiled with? Because you could just do this in C99:

```suggestion
        EdbStmtInfoSemState state = {0};
```

Reply:
Yeah, mostly assuming C99 in #7869 now